File: lab01/builder/src/main/java/com/example/builder/ArcherBuilder.java
package com.example.builder;

import com.example.builder.enums.CharacterType;
import com.example.builder.enums.WeaponType;

public class ArcherBuilder implements CharacterBuilder {

    private String name;
    private Integer health;
    private Integer mana;

    @Override
    public CharacterBuilder reset() {
        name = null;
        health = null;
        mana = null;
        return this;
    }

    @Override
    public CharacterBuilder setName(String name) {
        this.name = name;
        return this;
    }

    @Override
    public CharacterBuilder setHealth(Integer health) {
        this.health = health;
        return this;
    }

    @Override
    public CharacterBuilder setMana(Integer mana) {
        this.mana = mana;
        return this;
    }

    @Override
    public CharacterBuilder setWeapon(WeaponType weapon) {
        return this;
    }

    @Override
    public Character build() {
        return new Character(CharacterType.ARCHER, name, health, mana, null);
    }
}

File: lab01/builder1/Builders/ArcherBuilder.cs
using Builder1.Enums;
using Builder1.Models;

namespace Builder1.Builders;

public class ArcherBuilder : ICharacterBuilder
{
    private string? _name;
    private int? _health;
    private int? _mana;

    public ICharacterBuilder Reset()
    {
        _name = null;
        _health = null;
        _mana = null;
        return this;
    }

    public ICharacterBuilder SetName(string name) { _name = name; return this; }
    public ICharacterBuilder SetHealth(int? health) { _health = health; return this; }
    public ICharacterBuilder SetMana(int? mana) { _mana = mana; return this; }
    public ICharacterBuilder SetWeapon(WeaponType? weapon) => this; // Archer does not use weapons

    public Character Build() => new(CharacterType.Archer, _name!, _health, _mana, null);
}
